From 943da263b607d1751cf803956ffb1b152a7ac748 Mon Sep 17 00:00:00 2001
From: Zhuang Jin Can <jin.can.zhuang@intel.com>
Date: Tue, 19 May 2015 11:10:33 +0800
Subject: [PATCH] REVERTME: WORKAROUND: fix the long training sequences

This is a temporary workaround to fix the long training sequences issue
found in SSIC. The patch must be reverted once the clean solution is
integrated in BIOS.

Change-Id: I5b5b1d27b5499cc74fbb8691be0bd9c3fe15aac5
Signed-off-by: Zhuang Jin Can <jin.can.zhuang@intel.com>
---
 drivers/usb/host/xhci-pci.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/usb/host/xhci-pci.c b/drivers/usb/host/xhci-pci.c
index c48a48108d00..a8cc0ad19475 100644
--- a/drivers/usb/host/xhci-pci.c
+++ b/drivers/usb/host/xhci-pci.c
@@ -41,6 +41,9 @@
 
 #define PCI_DEVICE_ID_INTEL_CHT_XHCI	0x22b5
 
+#define SSIC_SS_PORT_LINK_CTRL 0x80ec
+#define SSIC_SS_PORT_LINK_CTRL_U3_MASK (0x7 << 9)
+
 static const char hcd_name[] = "xhci_hcd";
 
 /* called after powerup, by probe or system-pm "wakeup" */
@@ -58,6 +61,17 @@ static int xhci_pci_reinit(struct xhci_hcd *xhci, struct pci_dev *pdev)
 			xhci_dbg(xhci, "Intel Vendor Capability init done\n");
 	}
 
+	if (pdev->vendor == PCI_VENDOR_ID_INTEL &&
+				pdev->device == PCI_DEVICE_ID_INTEL_CHT_XHCI) {
+		struct usb_hcd *hcd = xhci_to_hcd(xhci);
+		u32 data;
+
+		/* Clear bit 11:9 of Superspeed Port Link Control reg */
+		data = readl(hcd->regs + SSIC_SS_PORT_LINK_CTRL);
+		data &= ~SSIC_SS_PORT_LINK_CTRL_U3_MASK;
+		writel(data, hcd->regs + SSIC_SS_PORT_LINK_CTRL);
+	}
+
 	/* PCI Memory-Write-Invalidate cycle support is optional (uncommon) */
 	if (!pci_set_mwi(pdev))
 		xhci_dbg(xhci, "MWI active\n");
-- 
1.9.1

